File: src/my_py_turtle/my_py_turtle/turtle_controller.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
import random
import math
import time
from turtlesim.srv import Spawn, Kill
from turtlesim.msg import Pose
from geometry_msgs.msg import Twist
from example_interfaces.msg import Int64, String
from my_robot_interfaces.srv import CatchTurtle, SpawnTurtle
from my_robot_interfaces.msg import Turtle, TurtleArray


class TurtleControllerNode(Node):
    def __init__(self):
        super().__init__("turtle_controller")

        self.declare_parameter("max_alive_turtles", 9)
        self.declare_parameter("kill_distance_threshold", 0.6)
        self.declare_parameter("linear_vel", 2.0)
        self.declare_parameter("radius", 0.5)
        self.declare_parameter("angle_decrement", 0.02)
        self.declare_parameter("p_gain", 1.5)
        self.declare_parameter("angular_gain", 2.0)

        self.counter_ = 2
        self.max_iterations = self.get_parameter("max_alive_turtles").value
        self.current_iteration = 0
        self.alive_turtles = {}
        self.kill_distance_threshold = self.get_parameter("kill_distance_threshold").value

        self.pose_ = None  # To store turtle1's current position so the timer can access it

        self.linear_vel = self.get_parameter("linear_vel").value
        self.radius = self.get_parameter("radius").value
        self.angle_decrement = self.get_parameter("angle_decrement").value

        self.p_gain = self.get_parameter("p_gain").value
        self.angular_gain = self.get_parameter("angular_gain").value
        

        self.spawn_client_ = self.create_client(Spawn, "trigger_spawn")
        self.kill_client_ = self.create_client(Kill, "trigger_kill")
        self.myspawn_client_ = self.create_client(SpawnTurtle, "trigger_myspawn")
        self.catch_client_ = self.create_client(CatchTurtle, "trigger_catch")

        while not self.spawn_client_.wait_for_service(timeout_sec=1.0):
            self.get_logger().info("Waiting for trigger_spawn service...")

        while not self.kill_client_.wait_for_service(timeout_sec=1.0):
            self.get_logger().info("Waiting for trigger_kill service...")

        while not self.myspawn_client_.wait_for_service(timeout_sec=1.0):
            self.get_logger().info("Waiting for trigger_myspawn service...")

        while not self.catch_client_.wait_for_service(timeout_sec=1.0):
            self.get_logger().info("Waiting for trigger_catch service...")


        # Create subscriber
        self.turtles_subscriber_ = self.create_subscription(Turtle, "alive_turtles", 
                                                            self.callback_alive_turtles, 10)
        self.turtles_array_subscriber_ = self.create_subscription(TurtleArray, "turtles_array",
                                                            self.callback_turtles_array, 10)
        self.pose_subscriber_ = self.create_subscription(Pose, "turtle1/pose", 
                                                         self.pose_callback, 10)


        self.pose_publisher_ = self.create_publisher(String, "turtle1_pose", 10)
        self.cmd_vel_publisher_ = self.create_publisher(Twist, "turtle1/cmd_vel", 10)
        self.turtles_array_publisher_ = self.create_publisher(TurtleArray, "turtles_array", 10)

        self.timer_ = self.create_timer(0.25, self.control_loop)

        self.move_timer_ = self.create_timer(0.1, self.move_to_closest_turtle)

        # Logging
        self.get_logger().info("Turtle Controller Node has started.")


    def pose_callback(self, msg: Pose):
        # self.get_logger().info(f"Turtle pose: ({str(msg.x)} , {str(msg.y)}) with {str(msg.theta)} degrees")
        # self.get_logger().info(str(self.alive_turtles))
        self.pose_ = msg
        pose_message = String()
        pose_message.data = f"Turtle pose: ({str(msg.x)} , {str(msg.y)}) with {str(msg.theta)} degrees"
        self.pose_publisher_.publish(pose_message)
        t1_x = msg.x
        t1_y = msg.y
        # Eating turtle logic
        for name in list(self.alive_turtles.keys()):
            pos = self.alive_turtles[name]
            dist = math.sqrt((t1_x - pos[0])**2 + (t1_y - pos[1])**2)
            if dist < self.kill_distance_threshold:
                self.kill_turtle(name)
                # self.spawn_new_turtle()
                self.myspawn_new_turtle()

    def move_spiral(self):
        cmd = Twist()
        # Constant forward motion
        cmd.linear.x = self.linear_vel
        # Angular velocity decreases over time to expand the spiral
        # msg.angular.z = v / r
        cmd.angular.z = self.linear_vel / self.radius
        self.radius += 0.01
        self.cmd_vel_publisher_.publish(cmd)

    def move_to_closest_turtle(self):
        if self.pose_ is None:
            return
        if not self.alive_turtles:
            # self.move_spiral()
            self.get_logger().info("No turtles to hunt?")
            return
        
        # Find the closest turtle
        t1_x = self.pose_.x
        t1_y = self.pose_.y
        closest_name = None
        min_dist = math.inf
        target_x, target_y = 0.0, 0.0

        for name, pos in self.alive_turtles.items():
            dist = math.sqrt((t1_x - pos[0])**2 + (t1_y - pos[1])**2)
            if dist < min_dist:
                target_x, target_y = pos[0], pos[1]
                min_dist = dist
                closest_name = name
        
        # Calculate the motion
        target_angle = math.atan2(target_y - t1_y, target_x - t1_x)
        steering_angle = target_angle - self.pose_.theta

        # Normalize angle_error to stay between -pi and pi (prevents spinning 350 degrees)
        steering_angle = math.atan2(math.sin(steering_angle), math.cos(steering_angle))

        # Generating the Twist command
        cmd = Twist()
        if min_dist > 0.1:  # Tolerance: stop if we are 0.1 units away
            # Proportional Control: move faster if far, slow down as you get closer
            cmd.linear.x = self.p_gain * min_dist  
            cmd.angular.z = self.angular_gain * steering_angle
        else:
            # Goal reached, Stop
            cmd.linear.x = 0.0
            cmd.angular.z = 0.0
        self.cmd_vel_publisher_.publish(cmd)


    def control_loop(self):
        if self.current_iteration >= self.max_iterations:
            self.timer_.cancel()
            return
        
        # self.spawn_new_turtle()
        self.myspawn_new_turtle()
        # create one-shot timer
        self.kill_timer_ = self.create_timer(
            0.75,
            self.kill_timer_callback
        )
        self.current_iteration += 1



    def kill_timer_callback(self):
        self.kill_timer_.cancel()
        self.kill_last_turtle()


    def callback_alive_turtles(self, msg: Turtle):
        self.get_logger().info("Received new turtle")
        self.alive_turtles[msg.name] = [msg.x, msg.y]
        self.publish_alive_turtles_array()


    def callback_turtles_array(self, msg: TurtleArray):
        # Clear the alive turtles dictionary
        self.alive_turtles.clear()
        # rebuild the dictionary from message
        for turtle in msg.turtles:
            self.alive_turtles[turtle.name] = [turtle.x, turtle.y]

        self.get_logger().info(f"Updated alive turtles dictionary: {len(self.alive_turtles)} turtles")
    

    def publish_alive_turtles_array(self):
        # Dictionary = source of truth
        # TurtleArray = snapshot copy

        msg = TurtleArray()
        for name, pos in self.alive_turtles.items():
            turtle = Turtle()
            turtle.name = name
            turtle.x = pos[0]
            turtle.y = pos[1]
            turtle.theta = 0.0
            msg.turtles.append(turtle)
        self.turtles_array_publisher_.publish(msg)


    def myspawn_new_turtle(self):
        request = SpawnTurtle.Request()
        request.x = random.uniform(1.0, 10.0)
        request.y = random.uniform(1.0, 10.0)
        request.theta = random.uniform(0.0, 6.28)
        request.name = 'turtle' + str(self.counter_)
        future = self.myspawn_client_.call_async(request)
        # self.alive_turtles[request.name] = [request.x, request.y]
        future.add_done_callback(self.callback_call_spawn)
        self.counter_ += 1
        self.last_spawned_name_ = None


    def spawn_new_turtle(self):

        request = Spawn.Request()
        # Turtlesim coordinates are roughly 0.0 to 11.0
        request.x = random.uniform(1.0, 10.0)
        request.y = random.uniform(1.0, 10.0)
        request.theta = random.uniform(0.0, 6.28)
        request.name = 'turtle' + str(self.counter_)

        future = self.spawn_client_.call_async(request)
        # self.alive_turtles[request.name] = [request.x, request.y]
        future.add_done_callback(self.callback_call_spawn)
        self.counter_ += 1
        self.last_spawned_name_ = None

    def kill_turtle(self, name):
        # request = Kill.Request()
        request = CatchTurtle.Request()
        request.name = name

        # future = self.kill_client_.call_async(request)
        future = self.catch_client_.call_async(request)

        future.add_done_callback(lambda future: self.after_kill(name))
        

    def kill_last_turtle(self):
        return
        # if self.last_spawned_name_ is None:
        #     return
        # request = Kill.Request()
        # request.name = self.last_spawned_name_
        # future = self.kill_client_.call_async(request)
        # future.add_done_callback(self.callback_call_kill)

    def after_kill(self, name):
        if name in self.alive_turtles:
            del self.alive_turtles[name]
        self.publish_alive_turtles_array()
        self.get_logger().info(f"Ate {name}!")




    def callback_call_spawn(self, future):
        try:
            response = future.result()
            self.last_spawned_name_ = response.name
            self.get_logger().info(f"Successfully spawned: {response.name}")
        except Exception as e:
            self.get_logger().error(f"Service call failed: {e}")

    def callback_call_kill(self, future):
        try:
            response = future.result()
            self.get_logger().info(f"Successfully killed turtle")
        except Exception as e:
            self.get_logger().error(f"Service call failed: {e}")



def main(args=None):
    rclpy.init(args=args)
    node = TurtleControllerNode()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: src/my_cpp_turtle/src/turtle_controller.cpp
#include "rclcpp/rclcpp.hpp"
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/string.hpp"
#include "my_robot_interfaces/srv/catch_turtle.hpp"
#include "my_robot_interfaces/srv/spawn_turtle.hpp"
#include "my_robot_interfaces/msg/turtle.hpp"
#include "my_robot_interfaces/msg/turtle_array.hpp"
#include "turtlesim/srv/spawn.hpp"
#include "turtlesim/srv/kill.hpp"
#include "turtlesim/msg/pose.hpp"
#include "geometry_msgs/msg/twist.hpp"

#include <cmath>
#include <random>
#include <sstream>


using namespace std::placeholders;
using namespace std::chrono_literals;


class TurtleControllerNode : public rclcpp::Node
{
public:
    TurtleControllerNode() : Node("turtle_controller"), counter_(2)
    {
        // === Parameters ===
        this->declare_parameter("max_alive_turtles", 9);
        this->declare_parameter("kill_distance_threshold", 0.6);
        this->declare_parameter("linear_vel", 2.0);
        this->declare_parameter("radius", 0.5);
        this->declare_parameter("angle_decrement", 0.02);
        this->declare_parameter("p_gain", 1.5);
        this->declare_parameter("angular_gain", 2.0);

        max_iterations_ = this->get_parameter("max_alive_turtles").as_int();
        kill_distance_threshold_ = this->get_parameter("kill_distance_threshold").as_double();
        linear_vel_ = this->get_parameter("linear_vel").as_double();
        radius_ = this->get_parameter("radius").as_double();
        angle_decrement_ = this->get_parameter("angle_decrement").as_double();
        p_gain_ = this->get_parameter("p_gain").as_double();
        angular_gain_ = this->get_parameter("angular_gain").as_double();

        // random generator
        std::random_device rd;
        gen_.seed(rd());
        
        // === Clients ===

        spawn_client_ = this->create_client<turtlesim::srv::Spawn>("spawn");
        kill_client_ = this->create_client<turtlesim::srv::Kill>("kill");
        myspawn_client_ = this->create_client<my_robot_interfaces::srv::SpawnTurtle>("myspawn");
        catch_client_ = this->create_client<my_robot_interfaces::srv::CatchTurtle>("catch");

        while (!spawn_client_->wait_for_service(1s)) {
            if (!rclcpp::ok()) {
                RCLCPP_INFO(this->get_logger(), "Interrupted while waiting for /spawn service...");
                return;
                }
                RCLCPP_INFO(this->get_logger(), "/spawn service not available, waiting again...");
        }
        while (!kill_client_->wait_for_service(1s)) {
            if (!rclcpp::ok()) {
                RCLCPP_INFO(this->get_logger(), "Interrupted while waiting for /kill service...");
            return;
            }
                RCLCPP_INFO(this->get_logger(), "/kill service not available, waiting again...");
        }
        while (!catch_client_->wait_for_service(1s)) {
            if (!rclcpp::ok()) {
                RCLCPP_INFO(this->get_logger(), "Interrupted while waiting for /catch service...");
                return;
                }
                RCLCPP_INFO(this->get_logger(), "/catch service not available, waiting again...");
        }
        while (!myspawn_client_->wait_for_service(1s)) {
            if (!rclcpp::ok()) {
                RCLCPP_INFO(this->get_logger(), "Interrupted while waiting for /myspawn service...");
            return;
            }
                RCLCPP_INFO(this->get_logger(), "/myspawn service not available, waiting again...");
        }

        // === Publishers ===
        pose_publisher_ = this->create_publisher<example_interfaces::msg::String>("turtle1_pose",10);
        cmd_vel_publisher_ = this->create_publisher<geometry_msgs::msg::Twist>("turtle1/cmd_vel", 10);
        turtles_array_publisher_ = this->create_publisher<my_robot_interfaces::msg::TurtleArray>("turtles_array", 10);

        // Subscribers
        pose_subscriber_ = this->create_subscription<turtlesim::msg::Pose>("turtle1/pose",10,
        std::bind(&TurtleControllerNode::pose_callback, this, std::placeholders::_1));
        turtles_array_subscriber_ = this->create_subscription<my_robot_interfaces::msg::TurtleArray>("turtles_array",10,
        std::bind(&TurtleControllerNode::callback_turtles_array, this, std::placeholders::_1));
        turtles_subscriber_ = this->create_subscription<my_robot_interfaces::msg::Turtle>("alive_turtles",10,
        std::bind(&TurtleControllerNode::callback_alive_turtles, this, std::placeholders::_1));

        // === Timers ===
        timer_ = this->create_wall_timer(250ms, 
        std::bind(&TurtleControllerNode::control_loop, this));
        move_timer_ = this->create_wall_timer(100ms,
        std::bind(&TurtleControllerNode::move_to_closest_turtle, this));


        RCLCPP_INFO(this->get_logger(), "Turtle Controller Node has started.");
    }

private:
    void pose_callback(const turtlesim::msg::Pose::SharedPtr msg)
    {
        pose_ptr_ = msg;
        auto pose_message = example_interfaces::msg::String();
        std::stringstream ss;
        ss << "Turtle pose: (" << pose_ptr_->x << "," << pose_ptr_->y <<  ") with " 
                            << pose_ptr_->theta << "degrees";
        pose_message.data = ss.str();
        pose_publisher_->publish(pose_message);

        // Eating turtle logic
        auto it = alive_turtles_.begin();
        while (it != alive_turtles_.end()){
            double dx = msg->x - it->second.first;
            double dx = msg->y = it->second.second;
            double dist = std::sqrt(std::pow(dx,2) + std::pow(dx,2));

            if (dist < this->kill_distance_threshold_){
                this->kill_turtle(it->first);
                this->myspawn_new_turtle();
                it = alive_turtles_.erase(it);
            }
            else
            {
                ++it;
            }
        }
    }

    void move_spiral()
    {
        auto cmd = geometry_msgs::msg::Twist();
        cmd.linear.x = this->linear_vel_;
        cmd.angular.z = this->linear_vel_ / this->radius_;
        this->radius_ += 0.01;
        cmd_vel_publisher_->publish(cmd);
    }
    
    void move_to_closest_turtle()
    {
        //
    }
    
    void control_loop()
    {
        //
    }

    void kill_timer_callback()
    {
        //
    }

    void callback_alive_turtles()
    {
        //
    }

    void callback_turtles_array()
    {
        //
    }

    void publish_alive_turtles_array()
    {
        //
    }

    void myspawn_new_turtle()
    {
        std::uniform_real_distribution<double> dist_pos(1.0, 10.0);
        std::uniform_real_distribution<double> dist_theta(0.0, 6.28);
        // What it does: It allocates memory for the Request object and returns a SharedPtr pointing to it.
        auto request = std::make_shared<my_robot_interfaces::srv::SpawnTurtle::Request>();
        request->x = dist_pos(gen_);
        request->y = dist_pos(gen_);
        request->theta = dist_theta(gen_);
        request->name = "turtle" + std::to_string(this->counter_++);

        myspawn_client_->async_send_request(request,
                        std::bind(&TurtleControllerNode::callback_call_myspawn,this,_1));
        
        this->last_spawned_name = "";
    }

    void spawn_new_turtle()
    {
        std::uniform_real_distribution<double> dist_pos(1.0, 10.0);
        std::uniform_real_distribution<double> dist_theta(0.0, 6.28);
        auto request = std::make_shared<turtlesim::srv::Spawn::Request>();

        request->x = dist_pos(gen_);
        request->y = dist_pos(gen_);
        request->theta = dist_theta(gen_);
        request->name = "turtle" + std::to_string(this->counter_++);

        spawn_client_->async_send_request(request,
                std::bind(&TurtleControllerNode::callback_call_spawn, this, _1));
        
        this->last_spawned_name = "";
    }

    void kill_turtle(const std::string name)
    {
        // auto request = std::make_shared<turtlesim::srv::Kill::Request>();
        auto request = std::make_shared<my_robot_interfaces::srv::CatchTurtle::Request>();
        request->name = name;
        catch_client_->async_send_request(request,
                        std::bind(&TurtleControllerNode::callback_call_kill, this, _1));
    }

    void kill_last_turtle()
    {
        //
    }

    void after_kill()
    {
        //
    }

    void callback_call_spawn(rclcpp::Client<turtlesim::srv::Spawn>::SharedFuture future)
    {
        try {
            // here in C++, the response is the result of the future
            auto response = future.get();
            this->last_spawned_name = response->name;
            RCLCPP_INFO(this->get_logger(), "Successfully spawned: %s", response->name.c_str());
        }
        catch (const std::exception &e){
            RCLCPP_ERROR(this->get_logger(), "Service call failed: %s", e.what());
        }
    }

    void callback_call_myspawn(rclcpp::Client<my_robot_interfaces::srv::SpawnTurtle>::SharedFuture future)
    {
        try {
            // here in C++, the response is the result of the future
            auto response = future.get();
            this->last_spawned_name = response->name;
            RCLCPP_INFO(this->get_logger(), "Successfully spawned: %s", response->name.c_str());
        }
        catch (const std::exception &e){
            RCLCPP_ERROR(this->get_logger(), "Service call failed: %s", e.what());
        }
    }

    void callback_call_kill(rclcpp::Client<my_robot_interfaces::srv::CatchTurtle>::SharedFuture future)
    {
        try {
            auto response = future.get();
            RCLCPP_INFO(this->get_logger(), "Successfully killed turtle");
        }
        catch(const std::exception &e) {
            RCLCPP_ERROR(this->get_logger(), "Service call failed: %s", e.what());
        }
    }


    int counter_;
    int max_iterations_;
    int current_iteration_ = 0;
    double kill_distance_threshold_;
    double p_gain_, angular_gain_, linear_vel_, radius_, angle_decrement_;

    turtlesim::msg::Pose::SharedPtr pose_ptr_; // array/vector
    std::map<std::string, std::pair<double, double>> alive_turtles_;
    std::string last_spawned_name;

    std::mt19937 gen_; // shuffler

    rclcpp::Client<turtlesim::srv::Spawn>::SharedPtr spawn_client_;
    rclcpp::Client<turtlesim::srv::Kill>::SharedPtr kill_client_;
    rclcpp::Client<my_robot_interfaces::srv::SpawnTurtle>::SharedPtr myspawn_client_;
    rclcpp::Client<my_robot_interfaces::srv::CatchTurtle>::SharedPtr catch_client_;

    rclcpp::Subscription<my_robot_interfaces::msg::Turtle>::SharedPtr turtles_subscriber_;
    rclcpp::Subscription<my_robot_interfaces::msg::TurtleArray>::SharedPtr turtles_array_subscriber_;
    rclcpp::Subscription<turtlesim::msg::Pose>::SharedPtr pose_subscriber_;

    rclcpp::Publisher<example_interfaces::msg::String>::SharedPtr pose_publisher_;
    rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr cmd_vel_publisher_;
    rclcpp::Publisher<my_robot_interfaces::msg::TurtleArray>::SharedPtr turtles_array_publisher_;
    
    rclcpp::TimerBase::SharedPtr timer_;
    rclcpp::TimerBase::SharedPtr move_timer_;


};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<TurtleControllerNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}
